Add optional timeout argument for RpcAgent join()

H-Huang opened the pull request:
Differential Revision: D35825382

Changes:
- Adds timeout argument to RpcAgent.join()
- Update API.py to also include fix bug (missing timeout for signal)
- Change default shutdown timeout to 0 (no timeout). Existing functionality in _all_gather will remain the same and wait indefinitely for signal if no timeout is set for the function. New functionality has user specify timeout for both the signal and rpc calls.

rohan-varma reviewed (commented):
Actually, wondering if this will run into the same issue as my previous attempt here? https://github.com/pytorch/pytorch/pull/70278

cc @pritamdamania87

mrshenli reviewed (approved):
LGTM

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -216,7 +216,8 @@ def _all_gather(obj, worker_names=None, timeout=UNSET_RPC_TIMEOUT):
         states = _all_gather_sequence_id_to_states[sequence_id]
-    states.proceed_signal.wait()
+    timeout = timeout if timeout else None
+    states.proceed_signal.wait(timeout=timeout)
```
Hey @H-Huang, just to double check. Would I be correct that the timeout here will be 0 ad it will make this call wait indefinitely, so that behavior does not change?


If so, could you please add a PR summary (and add to diffs as well) explaining that. Thanks!

I saw the following code:

```
    if timeout == UNSET_RPC_TIMEOUT:
        timeout = get_rpc_timeout()
```

Would I be correct that `get_rpc_timeout()` will return 0 by default?

H-Huang replied:
Thanks, for noticing this, I needed to fix some logic in the current implementation. Right now `UNSET_RPC_TIMEOUT = -1`,and `get_rpc_timeout()` gets whatever timeout is set by the RPC agent (default is 180).

Since we want to have keep the default behavior, of signal blocking indefinitely and RPC calls should keep their existing timeout. I added logic for two timeouts `signal_timeout` and `rpc_timeout`. https://github.com/pytorch/pytorch/pull/76194/files#diff-7f4f89fbe16f5d3e5e385d2625716f83a2e39477ec6c5d4cc51419bc0028b4c1R205-R211.

So the logic is now:
1. if timeout = UNSET_RPC_TIMEOU: RPC calls will default to the agent timeout, signal timeout will be infinite. This is the existing behavior.
2. if timeout is specified by user, RPC call and signal timeout will both use the same timeout. This is the added behavior.

mrshenli replied:
Makes sense to me! Thanks for the explanation!

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -343,7 +346,7 @@ def shutdown(graceful=True, timeout=DEFAULT_SHUTDOWN_TIMEOUT):
                 _wait_all_workers(timeout)
                 _delete_all_user_and_unforked_owner_rrefs()
-                agent.join(shutdown=True)
+                agent.join(shutdown=True, timeout=timeout)
```
lint error is real

H-Huang replied:
Updated `_distributed_rpc.pyi` to fix this

H-Huang committed:
Add optional timeout argument for RpcAgent join() (#76194)

Summary:
This PR was created to resolve issue brought up in https://fb.workplace.com/groups/319878845696681/permalink/741428653541696/

Changes:
- Adds timeout argument to RpcAgent.join()
- Add optional timeout argument to ThriftRpcAgent barrier()
- During shutdown (ThriftRpcAgent join) calls the barrier, the agent will use the timeout passed to shutdown and pass that timeout into the join().
- Update API.py to also include fix bug (missing timeout for signal)
- Change default shutdown timeout to 0 (no timeout). Existing functionality in _all_gather will remain the same and wait indefinitely for signal if no timeout is set for the function. New functionality has user specify timeout for both the signal and rpc calls.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76194

Test Plan:
Modified barrier test

buck test torch/fb/distributed/thriftRpcBackend/test:ThriftRpcAgentTest -- BarrierTest

Differential Revision: D35825382

fbshipit-source-id: 2195a4350c07accceec52905ef1f7990534d0ec8
